src/runtime/virtcontainers/device/config/config.go

```diff
@@ -270,6 +288,91 @@ type VFIODev struct {
 	IsPCIe bool
 }
 
+func (d VFIOPCIDev) GetID() *string {
+	return &d.ID
+}
+
+func (d VFIOPCIDev) GetType() *VFIODeviceType {
+	return &d.Type
+}
+
+func (d VFIOPCIDev) GetSysfsDev() *string {
+	return &d.SysfsDev
+}
+
+func (d VFIOPCIDev) GetBDF() *string {
+	return &d.BDF
+}
+
+func (d VFIOPCIDev) GetVendorID() *string {
+	return &d.VendorID
+}
+
+func (d VFIOPCIDev) GetDeviceID() *string {
+	return &d.DeviceID
+}
+
+func (d VFIOPCIDev) GetBus() *string {
+	return &d.Bus
+}
+
+func (d VFIOPCIDev) SetBus(bus string) {
+	d.Bus = bus
+}
+
+func (d VFIOPCIDev) GetGuestPciPath() *vcTypes.PciPath {
+	return &d.GuestPciPath
+}
+
+func (d VFIOPCIDev) SetGuestPciPath(guestPciPath vcTypes.PciPath) {
+	d.GuestPciPath = guestPciPath
+}
+
+func (d VFIOPCIDev) GetIsPCIe() *bool {
+	return &d.IsPCIe
+}
+
+func (d VFIOPCIDev) GetAPDevices() *[]string { return &[]string{} }
+
+type VFIOAPDev struct {
+	// ID is used to identify this drive in the hypervisor options.
+	ID string
+
+	// sysfsdev of VFIO mediated device
+	SysfsDev string
+
+	// Type of VFIO device
+	Type VFIODeviceType
+
+	// APDevices are the Adjunct Processor devices assigned to the mdev
+	APDevices []string
```

Comment:
Right, but AFAICT the mediated VFIO device should map to the `VFIODevice` which can already list multiple `VFIODev`s, each oone can be a single APQN.  This doesn't seem different to me from the PCI case where a single `/dev/vfio/NNN` (a single `VFIODevice`) corresponds to multiple hardware devices in the VFIO (each a `VFIODev`).  This can happen, for example, if you have a bunch of devices beneath a PCIe to PCI bridge.

Reply:
We go through `/sys/kernel/iommu_groups/NNN/devices`, but with VFIO-AP, that is just going to be one UUID, potentially with several APQNs behind it (see also https://docs.kernel.org/s390/vfio-ap.html). I think what you're describing applies to PCI normal, but not mediated.